das/state.go

```diff
@@ -42,34 +52,38 @@ func newCoordinatorState(params Parameters) coordinatorState {
 func (s *coordinatorState) resumeFromCheckpoint(c checkpoint) {
 	s.next = c.SampleFrom
 	s.networkHead = c.NetworkHead
-	// store failed to retry them on restart
-	for h, count := range c.Failed {
-		s.failed[h] = count
-		s.retry = append(s.retry, s.newJob(h, h))
+
+	for h := range c.Failed {
+		// TODO(@walldiss): reset retry counter to allow retries after restart. Will be removed when retry
+		// backoff is implemented.
+		s.failed[h] = 0
 	}
 }
 
 func (s *coordinatorState) handleResult(res result) {
 	delete(s.inProgress, res.id)
 
-	failedFromWorker := make(map[uint64]bool)
-	for _, h := range res.failed {
-		failedFromWorker[h] = true
-	}
-
 	// check if the worker retried any of the previously failed heights
 	for h := range s.failed {
 		if h < res.From || h > res.To {
 			continue
 		}
 
-		if !failedFromWorker[h] {
+		if res.failed[h] == 0 {
 			delete(s.failed, h)
 		}
 	}
-	// add newly failed heights
-	for h := range failedFromWorker {
-		s.failed[h]++
+
+	// update failed heights
+	for h := range res.failed {
+		failCount := 1
+		if res.job.Type == retryJob {
+			// if job was already in retry and failed again, persist attempt count
+			failCount += s.inRetry[h]
+			delete(s.inRetry, h)
+		}
+
+		s.failed[h] = failCount
```

Comment:
It's a bit confusing here - if the job was already retried once before, it should already be in the `s.failed` map so then just increment that value for the failed height by one, no? (this is in ref to my other comment regarding storing num retry attempts in result)

Reply:
When `retryJob` is created, failed height moved from failed map to inRetry map, so next job won't pick same height. If the job fails again, retry count needs to be restored from `inRetry` map and also be incremented by failed count that happen inside job (currently always 1).

For some reason I cannot answer to another comment about num of retry, so will post answer here:
I've moved from slice to map for fast lookup in the map. the failed count could only be 1 at this point, but it is much so 
 much nicer to work with actual fail count then checking and filling empty structs in multiple places, so I've sacrificed 4 bytes for that.